### ISSUE TITLE
`Ibc` and `ibc-proto` should have an std feature

### ISSUE THREAD

**Issue by Wizdave97** [informalsystems/ibc-rs#2015]
## Summary

ibc need to have an `std` feature to be compiled successfully in susbtrate 
Substrate compiles to both native and wasm and so we need std features enabled for dependencies for native compilation.
Also there are downstream dependencies that are pulling std into no_std environment in tendermint-rs
## Problem Definition
Both std feature and no_std features are required for compilation in substrate

## Proposal

I've created  an issue to track no_std in tendermint-rs here  https://github.com/informalsystems/tendermint-rs/issues/1108

I have a branch that implements this based on the change in this PR https://github.com/informalsystems/tendermint-rs/pull/1109.
But I have a blocker in fixing some of the functions here because the Event definition has changed, https://github.com/informalsystems/ibc-rs/blob/master/relayer/src/event/rpc.rs

## Acceptance Criteria

Compiles in both std and no_std in substrate

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate milestone (priority) applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by Wizdave97** [comment]
@romac this has been an issue with us compiling ibc-rs in  substrate, I've fixed `ibc` and `ibc-proto` and they compile in substrate with some changes I applied to tendermint-rs mentioned above, however I don't understand some of the relayer code that needs to be changed for the relayer crate to compile, but this is urgent for us, I would love to work to get both PRs merged by next week, so we can continue work on our substrate runtime.

**Comment by romac** [comment]
We do have such a feature for both crates.

**Comment by romac** [comment]
ibc-proto: https://github.com/informalsystems/ibc-rs/blob/29877ed44cefc367a81a8c7261907aac76849ca2/proto/Cargo.toml#L40

ibc: https://github.com/informalsystems/ibc-rs/blob/29877ed44cefc367a81a8c7261907aac76849ca2/modules/Cargo.toml#L21

**Comment by romac** [comment]
We also run a check for no-std compat for the ibc crate on CI: https://github.com/informalsystems/ibc-rs/tree/master/ci/no-std-check

**Comment by romac** [comment]
It would be awesome if you could help us augment our no-std check so that we catch the kind of incompatibility that you are running into so that we can then work towards better support for no-std based on your open PR. Thanks a lot for the latter by the way!

**Comment by Wizdave97** [comment]
> ibc-proto:
> 
> https://github.com/informalsystems/ibc-rs/blob/29877ed44cefc367a81a8c7261907aac76849ca2/proto/Cargo.toml#L40
> 
> ibc:
> 
> https://github.com/informalsystems/ibc-rs/blob/29877ed44cefc367a81a8c7261907aac76849ca2/modules/Cargo.toml#L21

Yeah, but the thing about substrate is that, in std mode, std must be enabled for all downstream dependencies  that have std features, hence the reason for the modification, although one of the  main issues was with tendermint-proto, the tendermint-rpc crate was a dependency of the tendermint-light-client-verifier and the tendermint-config which is an std crate was a dependency for the tendermint-rpc crate, so that was causing issues.


These kind of issues happen sporadically in susbtrate especially when there are other modules that have the same without std enabled in some dependencies, that's why it's best std is enabled completely for all dependencies in std mode

**Comment by soareschen** [comment]
> Yeah, but the thing about substrate is that, in std mode, std must be enabled for all downstream dependencies that have std features, hence the reason for the modification, although one of the main issues was with`ibc`.
> 
> These kind of issues happen sporadically in susbtrate especially when there are other modules that have the same without std enabled in some dependencies, that's why it's best std is enabled completely for all dependencies in std mode

- Can you elaborate more on what is the rationale of all `std` features of all crates must be enabled when Substrate is built in `std` mode?
- What is the purpose of `std` mode in Substrate? My understanding is that the whole reason we are working so much to support no_std is because Substrate does not support std.
- no_std is a subset of std, not an incompatible switch. I do not see any reason why no_std code cannot be used in std mode. If you run crates that support no_std like `ibc` with `--no-default-features`, you can see that the code in fact compiles even though `cargo build` always have `std` enabled.
- The way Cargo feature flag works is that as soon as any part of the dependency tree activates a feature like `std`, then it is activated globally for _all_ crates that use the dependency. For example if in your Substrate project you use a `foo` crate with `prost/std` enabled, then `prost/std` is enabled even for `ibc`. You can also use this quirk to include all the dependencies of interest at the top level and activate the `std` feature there.
- Substrate is just one of many Rust projects. I think it is unreasonable to expect all Rust projects to adopt a specific convention dictated by Substrate for the sake of making it work on Substrate. If Substrate needs crate to support no_std, they should provide guidelines such that Rust projects can check for no_std compliance _without_ having to build their projects with Substrate. We have to come up with our own `no-std-check` crate because Substrate do not provide any such support at the moment. If our approach `no-std-check` is not sufficient, please refer to this issue and ask Substrate to provide a better solution that does not involve setting up the entire Substrate toolchain.

> the `tendermint-rpc` crate was a dependency of the `tendermint-light-client-verifier` even though it was unused and the tendermint-config which is an std crate was a dependency for the `tendermint-rpc` crate, so that was causing some issues.

This was a dangling dependency that should be removed in the next tendermint-rs release, which I have just requested to be released soon. For now, you can add the following section in Cargo.toml for no_std support in tendermint-rs:

```toml
[patch.crates-io]
tendermint                        = { git = "https://github.com/informalsystems/tendermint-rs", branch = "v0.23.x" }
tendermint-proto                  = { git = "https://github.com/informalsystems/tendermint-rs", branch = "v0.23.x" }
tendermint-light-client-verifier  = { git = "https://github.com/informalsystems/tendermint-rs", branch = "v0.23.x" }
```

**Comment by Wizdave97** [comment]
> > Yeah, but the thing about substrate is that, in std mode, std must be enabled for all downstream dependencies that have std features, hence the reason for the modification, although one of the main issues was with`ibc`.
> > These kind of issues happen sporadically in susbtrate especially when there are other modules that have the same without std enabled in some dependencies, that's why it's best std is enabled completely for all dependencies in std mode
> 
> * Can you elaborate more on what is the rationale of all `std` features of all crates must be enabled when Substrate is built in `std` mode?
> * What is the purpose of `std` mode in Substrate? My understanding is that the whole reason we are working so much to support no_std is because Substrate does not support std.
> * no_std is a subset of std, not an incompatible switch. I do not see any reason why no_std code cannot be used in std mode. If you run crates that support no_std like `ibc` with `--no-default-features`, you can see that the code in fact compiles even though `cargo build` always have `std` enabled.
> 
> > These kind of issues happen sporadically in susbtrate especially when there are other modules that have the same without std enabled in some dependencies, that's why it's best std is enabled completely for all dependencies in std mode
> 
> * The `std` flag is a Cargo feature flag just like any other feature flags. It does not have any meaning on its own, and the absence of `std` flag does not prevent the `std` crate from being used.
> * The way Cargo feature flag works is that as soon as any part of the dependency tree activates a feature like `std`, then it is activated globally for _all_ crates that use the dependency. For example if in your Substrate project you use a `foo` crate with `prost/std` enabled, then `prost/std` is enabled even for `ibc`. You can also use this quirk to include all the dependencies of interest at the top level and activate the `std` feature there.
> * Substrate is just one of many Rust projects. I think it is unreasonable to expect all Rust projects to adopt a specific convention dictated by Substrate for the sake of making it work on Substrate. If Substrate needs crate to support no_std, they should provide guidelines such that Rust projects can check for no_std compliance _without_ having to build their projects with Substrate. We have to come up with our own `no-std-check` crate because Substrate do not provide any such support at the moment. If our approach `no-std-check` is not sufficient, please refer to this issue and ask Substrate to provide a better solution that does not involve setting up the entire Substrate toolchain.
> 
> > the `tendermint-rpc` crate was a dependency of the `tendermint-light-client-verifier` even though it was unused and the tendermint-config which is an std crate was a dependency for the `tendermint-rpc` crate, so that was causing some issues.
> 
> This was a dangling dependency that should be removed in the next tendermint-rs release, which I have just requested to be released soon. For now, you can add the following section in Cargo.toml for no_std support in tendermint-rs:
> 
> ```toml
> [patch.crates-io]
> tendermint                        = { git = "https://github.com/informalsystems/tendermint-rs", branch = "v0.23.x" }
> tendermint-proto                  = { git = "https://github.com/informalsystems/tendermint-rs", branch = "v0.23.x" }
> tendermint-light-client-verifier  = { git = "https://github.com/informalsystems/tendermint-rs", branch = "v0.23.x" }
> ```

Thanks for this, the patch fixes the compile issues.

**Comment by seunlanlege** [comment]
> Can you elaborate more on what is the rationale of all std features of all crates must be enabled when Substrate is built in std mode?

substrate runtime modules are compiled in both `std` and `no_std` mode, this is because if the onchain wasm code and native code have the same version, native code can be used to process incoming blocks to achieve faster block times.
